### ISSUE TITLE
[explicit-function-return-type] `allowTypedFunctionExpressions` option should be aware of JSX

### ISSUE THREAD

**Issue by merrywhether** [typescript-eslint/typescript-eslint#7552]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

The [`explicit-function-return-type` rule](https://typescript-eslint.io/rules/explicit-function-return-type/) has the [`allowTypedFunctionExpressions` option](https://typescript-eslint.io/rules/explicit-function-return-type/#allowtypedfunctionexpressions). This option is enabled as part of the [default option set for the rule](https://typescript-eslint.io/rules/explicit-function-return-type/#options).

This option is meant to allow omitting a return type when it will be enforced during assignment against a LHS type assertion, similar to how omitted argument types will be handled by TypeScript itself. This option should take into account assignment to JSX attributes but does not. Such assignment will be fully enforced by TypeScript JSX handling and is semantically the same as the object-attribute-assignment logic that is also allowed under this option.

<img width="638" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/4097742/0b0567fd-1727-41b9-b49e-210d39bfaad9">

As seen in this screenshot [from the linked repo](https://github.com/merrywhether/return-type/blob/main/src/pages/index.tsx), examples 2 and 3 are functionally the same but treated differently by the rule. Example 3 should not be a lint error.

[Demo PR](https://github.com/merrywhether/typescript-eslint/pull/2) with a potential version of this change.

### Reproduction Repository Link

https://github.com/merrywhether/return-type

### Repro Steps

1) clone the repo
2) `pnpm install`
3) `pnpm lint`

(other package managers will work as well)

```sh
% pn lint

> return-type@0.1.0 lint .../return-type
> next lint


./src/pages/index.tsx
15:5  Error: Missing return type on function.  @typescript-eslint/explicit-function-return-type
33:22  Error: Missing return type on function.  @typescript-eslint/explicit-function-return-type
```

(only the first error should be valid)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.5.0` |
| `@typescript-eslint/parser`            | `6.5.0` |
| `TypeScript`                           | `5.2.2` |
| `ESLint`                               | `8.48.0` |
| `node`                                 | `20.3.1` |

**Comment by merrywhether** [comment]
Looking through the code to size up a possible fix, using [this playground link](https://typescript-eslint.io/play/#ts=5.2.1-rc&showAST=es&fileType=.tsx&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAQAeO8%2Bc5AZgK4L5JwD2CZhCEjaF%2B2PKgBcYAExgQwAFDBgYACYQWAQzYwU%2BPizgBzEVu58wifDDYaoYLTBg8A7gBVcENQDEOXXgQAUXpBKChAqGkkQjYIZVUACyQkHCiVcQJiUgpYRCQAOl5gWJhoGnpGZjJ2Tgt%2BQWFRMkzgAGInF1dMtVqAvjocMIi%2BKEVFOABbHB5CFEywAG8wACE2FL4ACXcAWQAZAEEU4gAjdegAGjAAESEtODK1bf2ABUIeNKuAJQgtLgA5HgaMAAXzALHekzAAHJBH8kNCANzjBYAYT4RCEEDWGwQbw%2BDgAvDc7g9vM89vi0gAeRRgVbrJBbXaHY5wM5IaDUik4pnBMqTCDIAB8FzpYB5jL4QQFQqQimFyMUg1mKA02l0KD69TAIS00zKAAoAJTSH7wwHAxbilRgACMBTABkIgi4MCwYAQPBQiBYEBdWhOZXBs0czjcnjwvn89RCQ2gIwQY3pBiTKFcvwA1ry%2BFSiUtxfS%2BKjGJnpIaINIOJmva4EMawIThQX6a2naMeGUCi5jBXjci2yCxfSQUr6baZI7na6kO6rAg-QGgxAQ4Qw11I94-HVAnHhpFxamoPoMY1sVK8e80tJ0ZwzznLwTGy3W8XS%2BWIA2my%2B20fOxBux4XtPwHVsh3FUdxnpRoRAQMBDULMBqTUOAADdhUQ%2BlqQ5PklgKfCMy0bMLzzMFgAwtssJwyxFnwgpU0xTkH1IsByMwpDqIQdjx1UABmR1rFmGc5y9H0F39QhA2DFhQ06CMvGjHdpVCBMD0o18EBLOB8EzQlFj7Rtm2tdTfw7Lsez7UD1JBEF2LYttqWAFD0PFftFFsoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0URNGhjokcGAC%2BIHUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUgArEgA8KURJnQksAN3SQpAXxC6gA&tokens=false).

Would adding `JSXExpressionContainer` to the [conditional check in the `isValidFunctionExpressionReturnType` function](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts#L222-L228) be sufficient to handle this?

**Comment by merrywhether** [comment]
I've staged a PR against my fork (per repo rules) to test out an implementation.
https://github.com/merrywhether/typescript-eslint/pull/2
